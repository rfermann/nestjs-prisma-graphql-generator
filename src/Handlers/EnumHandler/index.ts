import type { Enum } from "../../types";
import { NestJSTypes } from "../../types";
import { BaseHandler } from "../BaseHandler";
import { compareObjectValues } from "../BaseHandler/compareFunctions";

export class EnumHandler extends BaseHandler {
  private readonly _enums: Enum[] = [];

  async createBarrelFile(): Promise<void> {
    if (this._enums.length) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/index`);

      sourceFile.addExportDeclarations(
        this._enums.map(({ name }) => {
          return {
            moduleSpecifier: `./${name}`,
            namedExports: [name],
          };
        })
      );
      await sourceFile.save();
    }
  }

  async createFiles(): Promise<void> {
    // eslint-disable-next-line no-restricted-syntax
    for (const { documentation, name: enumName, values } of this._enums) {
      const sourceFile = this.baseFileGenerator.createSourceFile(`${this.config.paths.enum}/${enumName}`);

      this.baseFileGenerator.addNestJSImports({ nestJSImports: [NestJSTypes.RegisterEnumType], sourceFile });

      sourceFile.addEnum({
        isExported: true,
        members: values.map((data) => {
          return {
            name: data.name,
            value: data.value,
          };
        }),
        name: enumName,
      });

      sourceFile.addStatements((writer) =>
        writer
          .writeLine("")
          .writeLine(`${NestJSTypes.RegisterEnumType}(${enumName}, {`)
          .writeLine(`name: "${enumName}",`)
          .conditionalWrite(typeof documentation === "string", `description: "${documentation}"`)
          .writeLine("});")
      );

      sourceFile.formatText();
    }

    await this.baseFileGenerator.save();
  }

  parse(): void {
    this.baseParser.dmmf.datamodel.enums.forEach(({ documentation, name, values }): void => {
      this._enums.push({
        documentation,
        name: this.baseParser.getEnumName(name),
        values: values.map(({ name: value }) => {
          return { name: value, value };
        }),
      });
    });

    this.baseParser.dmmf.schema.enumTypes.prisma.forEach(({ name, values }): void => {
      this._enums.push({
        name: this.baseParser.getEnumName(name),
        values: values.map((value) => {
          return { name: value, value };
        }),
      });
    });

    this._enums.sort((a, b) => compareObjectValues({ a, b, field: "name" }));
  }
}
