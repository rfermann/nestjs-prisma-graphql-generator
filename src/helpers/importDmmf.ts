/* eslint-disable global-require */
/* eslint-disable import/no-dynamic-require */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import type { DMMF } from "@prisma/generator-helper";

export const importDmmf = (prismaClientImportPath: string): DMMF.Document =>
  require(prismaClientImportPath).dmmf as DMMF.Document;
