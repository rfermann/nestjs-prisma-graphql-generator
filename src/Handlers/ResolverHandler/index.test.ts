// eslint-disable-next-line import/no-nodejs-modules
import { existsSync, readdirSync, readFileSync, rmdirSync } from "fs";

import type { GeneratorConfig } from "../../GeneratorConfig";
import { importDmmf } from "../../helpers";
import { EnumHandler } from "../EnumHandler";

import { ResolverHandler } from ".";

describe("ResolverHandler", () => {
  // eslint-disable-next-line jest/no-hooks
  beforeAll(async () => {
    if (existsSync(`${process.cwd()}/fixtures/resolverTypeHandler/fixtures/`)) {
      rmdirSync(`${process.cwd()}/fixtures/resolverTypeHandler/fixtures/`, { recursive: true });
    }
  });

  it("should parse resolvers and create correct resolver files from datamodel", async () => {
    expect.assertions(12);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/resolverTypeHandler/fixtures/output2`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      paths: {
        enums: "enums",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "PrismaService",
      prismaServiceImportPath: "@PrismaService",
    };

    const enumHandler = new EnumHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    const resolverTypeHandler = new ResolverHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    enumHandler.parse();

    resolverTypeHandler.parse(enumHandler.getEnums());
    await resolverTypeHandler.createFiles();

    const userResolverFiles = readdirSync(`${config.basePath}/User/${config.paths.resolvers}`);

    userResolverFiles.forEach((resolver) => {
      const userResolverFile = readFileSync(`${config.basePath}/User/${config.paths.resolvers}/${resolver}`, "utf-8");

      // eslint-disable-next-line jest/prefer-inline-snapshots
      expect(userResolverFile).toMatchSnapshot();
    });
    // eslint-disable-next-line jest/prefer-inline-snapshots
    expect(userResolverFiles).toMatchSnapshot();
  });

  it("should create correct barrel files", async () => {
    expect.assertions(2);

    const config: GeneratorConfig = {
      basePath: `${process.cwd()}/fixtures/resolverTypeHandler/fixtures/output3`,
      includePrismaSelect: false,
      inputArgumentsName: "",
      paths: {
        enums: "enums",
        inputTypes: "inputTypes",
        model: "model",
        outputTypes: "outputTypes",
        resolvers: "resolvers",
        shared: "shared",
      },
      prismaClientImportPath: `${process.cwd()}/node_modules/@prisma/client`,
      prismaServiceImport: "PrismaService",
      prismaServiceImportPath: "@PrismaService",
    };

    const resolverTypeHandler = new ResolverHandler({
      config,
      dmmf: importDmmf(config.prismaClientImportPath),
    });

    resolverTypeHandler.parse([]);
    await Promise.all([resolverTypeHandler.createBarrelFiles(), resolverTypeHandler.createFiles()]);

    const userResolverIndexFile = readFileSync(`${config.basePath}/User/${config.paths.resolvers}/index.ts`, "utf-8");
    const userResolverFiles = readdirSync(`${config.basePath}/User/${config.paths.resolvers}`);

    const lines = userResolverIndexFile
      .toString()
      .split("\n")
      .filter((line) => line.length).length;

    expect(lines).toBe(userResolverFiles.length - 1);
    expect(userResolverIndexFile).toMatchInlineSnapshot(`
      "export { AggregateUserResolver } from \\"./AggregateUserResolver\\";
      export { CreateManyUserResolver } from \\"./CreateManyUserResolver\\";
      export { CreateOneUserResolver } from \\"./CreateOneUserResolver\\";
      export { DeleteManyUserResolver } from \\"./DeleteManyUserResolver\\";
      export { DeleteOneUserResolver } from \\"./DeleteOneUserResolver\\";
      export { FindFirstUserResolver } from \\"./FindFirstUserResolver\\";
      export { FindManyUserResolver } from \\"./FindManyUserResolver\\";
      export { FindUniqueUserResolver } from \\"./FindUniqueUserResolver\\";
      export { UpdateManyUserResolver } from \\"./UpdateManyUserResolver\\";
      export { UpdateOneUserResolver } from \\"./UpdateOneUserResolver\\";
      export { UpsertOneUserResolver } from \\"./UpsertOneUserResolver\\";
      "
    `);
  });
});
