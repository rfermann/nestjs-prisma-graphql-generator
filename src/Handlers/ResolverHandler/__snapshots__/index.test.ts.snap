// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 1`] = `
"import { Args, Query, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { AggregateUser } from \\"../inputTypes/AggregateUser\\";
import { AggregateUser } from \\"../outputTypes/AggregateUser\\";
import { Prisma } from \\"@prisma/client\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class AggregateUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Query(() => AggregateUser, {
    nullable: false
  })
  async aggregateUser(@Args(\\"\\", { nullable: true }) : AggregateUser): Promise<Prisma.GetUserAggregateType<AggregateUser>> {
    return this._prismaService.user.aggregate({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 2`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { CreateManyUser } from \\"../inputTypes/CreateManyUser\\";
import { AffectedRowsOutput } from \\"../../shared/outputTypes/AffectedRowsOutput\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class CreateManyUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => AffectedRowsOutput, {
    nullable: false
  })
  async createManyUser(@Args(\\"\\", { nullable: true }) : CreateManyUser): Promise<AffectedRowsOutput | null> {
    return this._prismaService.user.createMany({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 3`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { CreateOneUser } from \\"../inputTypes/CreateOneUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class CreateOneUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => User, {
    nullable: false
  })
  async createOneUser(@Args(\\"\\", { nullable: true }) : CreateOneUser): Promise<User | null> {
    return this._prismaService.user.create({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 4`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { DeleteManyUser } from \\"../inputTypes/DeleteManyUser\\";
import { AffectedRowsOutput } from \\"../../shared/outputTypes/AffectedRowsOutput\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class DeleteManyUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => AffectedRowsOutput, {
    nullable: false
  })
  async deleteManyUser(@Args(\\"\\", { nullable: true }) : DeleteManyUser): Promise<AffectedRowsOutput | null> {
    return this._prismaService.user.deleteMany({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 5`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { DeleteOneUser } from \\"../inputTypes/DeleteOneUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class DeleteOneUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => User, {
    nullable: true
  })
  async deleteOneUser(@Args(\\"\\", { nullable: true }) : DeleteOneUser): Promise<User | null> {
    return this._prismaService.user.delete({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 6`] = `
"import { Args, Query, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { FindFirstUser } from \\"../inputTypes/FindFirstUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class FindFirstUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Query(() => User, {
    nullable: true
  })
  async findFirstUser(@Args(\\"\\", { nullable: true }) : FindFirstUser): Promise<User | null> {
    return this._prismaService.user.findFirst({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 7`] = `
"import { Args, Query, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { FindManyUser } from \\"../inputTypes/FindManyUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class FindManyUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Query(() => [User], {
    nullable: false
  })
  async findManyUser(@Args(\\"\\", { nullable: true }) : FindManyUser): Promise<User[] | null> {
    return this._prismaService.user.findMany({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 8`] = `
"import { Args, Query, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { FindUniqueUser } from \\"../inputTypes/FindUniqueUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class FindUniqueUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Query(() => User, {
    nullable: true
  })
  async findUniqueUser(@Args(\\"\\", { nullable: true }) : FindUniqueUser): Promise<User | null> {
    return this._prismaService.user.findUnique({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 9`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { UpdateManyUser } from \\"../inputTypes/UpdateManyUser\\";
import { AffectedRowsOutput } from \\"../../shared/outputTypes/AffectedRowsOutput\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class UpdateManyUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => AffectedRowsOutput, {
    nullable: false
  })
  async updateManyUser(@Args(\\"\\", { nullable: true }) : UpdateManyUser): Promise<AffectedRowsOutput | null> {
    return this._prismaService.user.updateMany({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 10`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { UpdateOneUser } from \\"../inputTypes/UpdateOneUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class UpdateOneUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => User, {
    nullable: true
  })
  async updateOneUser(@Args(\\"\\", { nullable: true }) : UpdateOneUser): Promise<User | null> {
    return this._prismaService.user.update({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 11`] = `
"import { Args, Mutation, Resolver } from \\"@nestjs/graphql\\";
import { } from \\"\\";
import { User } from \\"../../User/model\\";
import { UpsertOneUser } from \\"../inputTypes/UpsertOneUser\\";

@Resolver(() => User, {
  isAbstract: false,
})
export class UpsertOneUserResolver {
  private readonly _prismaService;

  constructor(prismaService) {
    this._prismaService = prismaService;
  }

  @Mutation(() => User, {
    nullable: false
  })
  async upsertOneUser(@Args(\\"\\", { nullable: true }) : UpsertOneUser): Promise<User | null> {
    return this._prismaService.user.upsert({ ... })
  }
}
"
`;

exports[`ResolverHandler should parse resolvers and create correct resolver files from datamodel 12`] = `
Array [
  "AggregateUserResolver.ts",
  "CreateManyUserResolver.ts",
  "CreateOneUserResolver.ts",
  "DeleteManyUserResolver.ts",
  "DeleteOneUserResolver.ts",
  "FindFirstUserResolver.ts",
  "FindManyUserResolver.ts",
  "FindUniqueUserResolver.ts",
  "UpdateManyUserResolver.ts",
  "UpdateOneUserResolver.ts",
  "UpsertOneUserResolver.ts",
]
`;
